### ISSUE TITLE
sklearn SVM model converted into ONNX model but gives different output results

### ISSUE THREAD

**Issue by totesarana** [microsoft/onnxruntime#11284]
I am using sklearn to train a SVM model and I converted it into ONNX model. However, when testing the two models with the same input, the outputs are different.

**System information**
- OS Platform and Distribution: Windows 10 Enterprise 64 bit
- ONNX Runtime installed from (source or binary): install using PIP from Python 3.9.7
- ONNX Runtime version: 1.11.0
- Python version: 3.9.7
- Visual Studio version (if applicable): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: Not using GPU

**To Reproduce**
`import pickle
from sklearn import svm
import numpy as np
from skl2onnx import convert_sklearn
from skl2onnx.common.data_types import FloatTensorType
import onnxruntime as rt
svm1_file = 'svm.model'
classifier_1 = pickle.load(open(svm1_file, 'rb'))
testingexample = [62, 62, 99, 99, 99, 99, 297, 297, 343, 343, 512, 0.0925, 262, 612, 322.6, 6.7963, 0.01993]
features = np.array([testingexample]).astype(np.float32)
result1_1 = classifier_1.predict(features)[0] #SKLEARN Result
result1_2 = classifier_1.predict_proba(features)[0]
print('svm1_file, SKLEARN result is %s, %s\n'%(str(result1_1), str(result1_2)))

#Convert into ONNX
initial_types = [('float_input',FloatTensorType([None, 17]))]
onx1 = convert_sklearn(classifier_1, initial_types = initial_types)
with open("Convert_svm.onnx", "wb") as f:
        f.write(onx1.SerializeToString())
sess1 = rt.InferenceSession(onx1.SerializeToString())
res1 = sess1.run(None, {'
[Onnx_and_sklearn_Models.zip](https://github.com/microsoft/onnxruntime/files/8525576/Onnx_and_sklearn_Models.zip)
float_input': features.astype(np.float32)})
print(res1) #ONNX result
`
Sklearn result  is 0.0, [**0.99006934 0.00993066**]
Onnx result is [array([0], dtype=int64), [{**0: 0.9864753484725952, 1: 0.01352467481046915**}]]

**Expected behavior**
The Sklearn mo
[Onnx_and_sklearn_Models.zip](https://github.com/microsoft/onnxruntime/files/8525578/Onnx_and_sklearn_Models.zip)
del result should match the ONNX model result.

Is there anything wrong in my code when converting the SVM learned by using SKLEARN into ONNX format?

**Comment by xadupre** [comment]
Hi @totesarana, the discrepancies are mostly due to the conversion from float to double. onnxruntime converts double to float before computing the probabilities in SVMClassifier and coefficients are converted into float when the model is converted. To reduce the discrepancies, ONNX specifications must be extended to support double coefficients (similar to this https://github.com/onnx/onnx/pull/3897) and onnxruntime must be updated to support this new specifications. Here is an experiment I ran with a python runtime. It shows that discrepancies are lower when the runtime computes the probabilities with double. Which version of onnxruntime did you use to compute the probabilities? The latest version of onnxruntime fails at loading the model.

```python
import pickle
import numpy
import onnx
from onnxruntime import InferenceSession
from mlprodict.plotting.text_plot import onnx_simple_text_plot
from mlprodict.onnx_tools.onnx_manipulations import select_model_inputs_outputs
from mlprodict.onnx_conv import to_onnx
from mlprodict.onnxrt import OnnxInference


skl_name = "Onnx_and_sklearn_Models/svm.model"
onx_name = "Onnx_and_sklearn_Models/Convert_svm.onnx"

with open(skl_name, "rb") as f:
    model = pickle.load(f)
with open(onx_name, "rb") as f:
    onx = onnx.load(f)

print("--ONNX--")
print(onnx_simple_text_plot(onx))

rnd = numpy.random.rand(5, 17).astype(numpy.float64) + 50
expected = model.predict_proba(rnd)
print("--EXPECTED--")
print(expected)

try:
    model1 = InferenceSession(onx.SerializeToString())
except Exception as e:
    print("--FAIL--", e)

onx2 = select_model_inputs_outputs(onx, outputs=['output_label', 'probabilities'])
model1 = InferenceSession(onx2.SerializeToString())
got = model1.run(None, {'double_input': rnd})[1]
print("--ORT--")
print(got)
print('--DIFF--')
diff = numpy.abs(got - expected)
print(diff)
print("MAX:", diff.max())

####################################
# With a runtime supported double for SVMClassifier

onx64 = to_onnx(model, rnd, options={'zipmap': False}, rewrite_ops=True)

print("--ONNX-64--")
print(onnx_simple_text_plot(onx64))

oinf = OnnxInference(onx64)
got64 = oinf.run({'X': rnd})['probabilities']
print("--PYRT--")
print(got64)
print('--DIFF--')
diff = numpy.abs(got64 - expected)
print(diff)
print("MAX:", diff.max())
```

output:

```
--ONNX--
opset: domain='ai.onnx.ml' version=1
opset: domain='' version=9
input: name='double_input' type=dtype('float64') shape=(0, 17)
SVMClassifier(double_input) -> label, SVM02
  Cast(label, to=7) -> output_label
Cast(SVM02, to=11) -> probabilities
  ZipMap(probabilities) -> output_probability
output: name='output_label' type=dtype('int64') shape=(0,)
output: name='output_probability' type='?' shape=()
--EXPECTED--
[[0.08900157 0.91099843]
 [0.09689244 0.90310756]
 [0.08190482 0.91809518]
 [0.08410505 0.91589495]
 [0.09315535 0.90684465]]
--FAIL-- [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. Type Error: Type 'tensor(double)' of input parameter (probabilities) of operator (ZipMap) in node (ZipMap) is invalid.
--ORT--
[[0.10351969 0.89648026]
 [0.09738901 0.9026109 ]
 [0.0982206  0.90177947]
 [0.08161405 0.91838592]
 [0.11160246 0.88839757]]
--DIFF--
[[0.01451812 0.01451817]
 [0.00049657 0.00049666]
 [0.01631578 0.01631571]
 [0.002491   0.00249097]
 [0.01844711 0.01844708]]
MAX: 0.0184471088385028
--ONNX-64--
opset: domain='mlprodict' version=1
opset: domain='' version=9
input: name='X' type=dtype('float64') shape=(0, 17)
SVMClassifierDouble[mlprodict](X) -> label, SVM02
  Cast(SVM02, to=11) -> probabilities
output: name='label' type=dtype('int64') shape=(0,)
output: name='probabilities' type=dtype('float64') shape=(0, 2)
--PYRT--
[[0.08987652 0.91012348]
 [0.09784254 0.90215746]
 [0.08271638 0.91728362]
 [0.08493528 0.91506472]
 [0.09407032 0.90592968]]
--DIFF--
[[0.00087495 0.00087495]
 [0.0009501  0.0009501 ]
 [0.00081156 0.00081156]
 [0.00083023 0.00083023]
 [0.00091497 0.00091497]]
MAX: 0.0009501018942673634
```

**Comment by xadupre** [comment]
It is the best today. It does not use onnxruntime but a custom runtime I developped to check that kind of situations and see whether or not it makes sense to bring a new operator into ONNX. I should be able to modify that runtime to reduce the discrepancies to 1e-5 after using double coefficients instead of floats. (1- 2 days) It would take 3,4 months to bring the same in onnx / onnxruntime (part of the release cycle).

About training, it is possible to train an ONNX model with onnxruntime-training but it assumes the model can be trained with a gradient algorithm. We could also write a custom training algorithm only with ONNX primitives. Both options require to have an existing implementation of a training algorithm with torch or numpy.

We could also to create a fork of scikit-learn and change type double into float in the code and recompile. But I don't see that kind of modifications be merged in the official scikit-learn library.

Last comment, you may want to remove operator ZipMap in the ONNX graph by using `options={'zipmap': False}`. The inference time is faster.

Maybe we could also use a neural network or a random forest using the onnx output and trying to fit the sklearn output to reduce the discrepancies.

Maybe there is another idea I did not see.

**Comment by totesarana** [comment]
Hi @xadupre, thanks for replying. Is there a reason why onnxruntime did not support double computing to begin with?

**Comment by xadupre** [comment]
SVMClassifier only allows float coefficients in the current version of onnx and onnxruntime just implements the same. We noticed significant discrepancies with non continuous functions (trees) and models with big matrix multiplications. Trees specifications was modified, matrix multiplication is already implementation with double. SVMClassifier is the last one. It is unclear whether we should enable double coefficient or try to decompose the operator into other ONNX primitives. I did not check if that was possible without introducing a new operator.
